ratis-netty/src/main/java/org/apache/ratis/netty/server/NettyServerStreamRpc.java

```diff
@@ -58,51 +63,55 @@
 
   private final StateMachine stateMachine;
   private final ConcurrentMap<Long, CompletableFuture<DataStream>> streams = new ConcurrentHashMap<>();
+  private final ConcurrentMap<Long, List<DataStreamOutput>> peersStreamOutput = new ConcurrentHashMap<>();
+
+  private List<DataStreamClient> clients = new ArrayList<>();
 
   public NettyServerStreamRpc(RaftPeer server, StateMachine stateMachine) {
     this.raftServer = server;
     this.stateMachine = stateMachine;
     this.channelFuture = buildChannel();
   }
 
-  private CompletableFuture<DataStream> getDataStreamFuture(ByteBuf buf, AtomicBoolean released) {
+  public NettyServerStreamRpc(
+      RaftPeer server, List<RaftPeer> otherPeers,
+      StateMachine stateMachine, RaftProperties properties){
+    this(server, stateMachine);
+    setupClient(otherPeers, properties);
+  }
+
+  private List<DataStreamOutput> getDataStreamOutput() {
+    return clients.stream().map(client -> client.stream()).collect(Collectors.toList());
```

Comment:
@szetszwo @amaliujia Should we use the streamId got from client to current server ? Maybe it's easy to debug when error happens. Not sure when the streamId will be different among 3 servers, but if streamId is different, the same stream has different streamId among 3 servers seems confused.

```
private List<DataStreamOutput> getDataStreamOutput(long streamId) {
    return clients.stream().map(client -> client.stream(streamId)).collect(Collectors.toList());
}
```

Reply:
This is a very good point. Right now indeed it only keeps a consistent mapping between stream ids. There might be a case that these stream id are different.

However, I am thinking even have a consistent stream ids for all servers, it might not be good enough for debugging. Ideally, we need a way to logging the path of a request, e.g. from client to server, then to other servers, and stream id mappings, and where is wrong. And then the best is aggregate such information and send back to the client. 


https://issues.apache.org/jira/browse/RATIS-1098 is partially related to this goal, meanwhile we might can extend RATIS-1098 or have another Jira to fully achieve this goal.